File: chrome/browser/password_manager/android/password_manager_lifecycle_helper.cc
// Copyright 2022 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/password_manager/android/password_manager_lifecycle_helper.h"

#include <jni.h>
#include <cstdint>
#include <utility>

#include "base/android/jni_android.h"
#include "chrome/browser/password_manager/android/jni_headers/PasswordManagerLifecycleHelper_jni.h"

PasswordManagerLifecycleHelper::PasswordManagerLifecycleHelper(
    base::RepeatingClosure foregrounding_callback) {
  java_object_ = Java_PasswordManagerLifecycleHelper_getInstance(
      base::android::AttachCurrentThread());
  foregrounding_callback_ = std::move(foregrounding_callback);
  Java_PasswordManagerLifecycleHelper_registerObserver(
      base::android::AttachCurrentThread(), java_object_,
      reinterpret_cast<intptr_t>(this));
}

PasswordManagerLifecycleHelper::~PasswordManagerLifecycleHelper() {
  Java_PasswordManagerLifecycleHelper_unregisterObserver(
      base::android::AttachCurrentThread(), java_object_,
      reinterpret_cast<intptr_t>(this));
}

void PasswordManagerLifecycleHelper::OnForegroundSessionStart(JNIEnv* env) {
  DCHECK(foregrounding_callback_);
  foregrounding_callback_.Run();
}

File: chrome/browser/password_manager/android/java/src/org/chromium/chrome/browser/password_manager/PasswordManagerLifecycleHelper.java
// Copyright 2022 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.password_manager;

import androidx.annotation.VisibleForTesting;

import org.jni_zero.CalledByNative;
import org.jni_zero.NativeMethods;

import org.chromium.base.ObserverList;

/** This helper forwards lifecycle events to password manager classes. */
public class PasswordManagerLifecycleHelper {
    private static PasswordManagerLifecycleHelper sInstance;
    private ObserverList<Long> mNativeObservers = new ObserverList<>();

    /**
     * Returns the singleton instance of this class and lazily creates it if that hasn't happened.
     *
     * @return The only {@link PasswordManagerLifecycleHelper}.
     */
    @CalledByNative
    public static PasswordManagerLifecycleHelper getInstance() {
        if (sInstance == null) sInstance = new PasswordManagerLifecycleHelper();
        return sInstance;
    }

    /** Notifies all observers that a foreground session has begun. */
    public void onStartForegroundSession() {
        for (Long observer : mNativeObservers) {
            assert observer != 0;
            PasswordManagerLifecycleHelperJni.get().onForegroundSessionStart(observer);
        }
    }

    @CalledByNative
    @VisibleForTesting
    void registerObserver(long nativePasswordManagerLifecycleHelper) {
        mNativeObservers.addObserver(nativePasswordManagerLifecycleHelper);
    }

    @CalledByNative
    @VisibleForTesting
    void unregisterObserver(long nativePasswordManagerLifecycleHelper) {
        mNativeObservers.removeObserver(nativePasswordManagerLifecycleHelper);
    }

    private PasswordManagerLifecycleHelper() {}

    /** C++ method signatures. */
    @NativeMethods
    interface Natives {
        void onForegroundSessionStart(long nativePasswordManagerLifecycleHelperImpl);
    }
}
